api/paidAction/index.js

```diff
@@ -95,271 +128,285 @@ export default async function performPaidAction (actionType, args, context) {
-}
-
-async function performPessimisticAction (actionType, args, context) {
-  const action = paidActions[actionType]
-
-  if (!action.supportsPessimism) {
-    throw new Error(`This action ${actionType} does not support pessimistic invoicing`)
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
+}
 
-  // just create the invoice and complete action when it's paid
-  const invoiceArgs = await createLightningInvoice(actionType, args, context)
-  return {
-    invoice: await createDbInvoice(actionType, args, context, invoiceArgs),
-    paymentMethod: 'PESSIMISTIC'
+async function performPessimisticAction (invoiceData, actionType, paidAction, args, context) {
+  const { models, actionId, me } = context
+  const performInvoicedAction = async tx => {
+    context.tx = tx
+    context.optimistic = false
+    await checkUser(me?.id ?? USER_ID.anon, context)
+
+    const invoiceEntry = await createDbInvoice(invoiceData, {
+      actionType,
+      actionId,
+      optimistic: false,
+      args
+    }, context)
+    return {
+      invoice: invoiceEntry,
+      paymentMethod: 'PESSIMISTIC',
+      retriable: invoiceEntry.retriable,
+      retriableWithFeeCredits: invoiceEntry.retriableWithFeeCredits,
+      retriableManually: invoiceEntry.retriableManually
+    }
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
 }
 
-export async function retryPaidAction (actionType, args, context) {
-  const { models, me } = context
-  const { invoice: failedInvoice } = args
+export async function retryPaidAction (actionType, { invoice, forceFeeCredits, forceRetry }, context) {
+  const { models } = context
+  const failedInvoice = invoice
 
-  console.log('retryPaidAction', actionType, args)
+  console.log('retryPaidAction', actionType, invoice)
 
-  const action = paidActions[actionType]
-  if (!action) {
+  const paidAction = paidActions[actionType]
+  if (!paidAction) {
     throw new Error(`retryPaidAction - invalid action type ${actionType}`)
   }
 
-  if (!me) {
-    throw new Error(`retryPaidAction - must be logged in ${actionType}`)
+  if (!failedInvoice) {
+    throw new Error(`retryPaidAction - missing invoice ${actionType}`)
   }
 
-  if (!action.supportsOptimism) {
-    throw new Error(`retryPaidAction - action does not support optimism ${actionType}`)
+  let retriable = false
+  if (forceRetry) {
+    retriable = true
+    context.retriedWithFeeCredits = false
+    context.retriedManually = false
+  } else {
+    if (failedInvoice.retriable) retriable = true
+    context.retriedWithFeeCredits = !failedInvoice.retriableWithFeeCredits
+    context.retriedManually = !failedInvoice.retriableManually
```

Comment:
should this be a single state with an enum? Maybe not, the 3 flags are not necessarily dependent to each other and we don't need to follow a specific flow.